### ISSUE TITLE
[Linux] FPS limit & Speed Scaler behavior, audio / video sync issues

### ISSUE THREAD

**Issue by lorn10** [grumpycoders/pcsx-redux#656]
Hi all!

I just want to note here that almost all "Soft GPU configuration" options in the Linux builds (like 02767acb) have been missing for quite some time. For me the most important is the **FPS limit** feature:
![pcsx-redux-missing_FPS](https://user-images.githubusercontent.com/9931771/138601182-064b697a-9db4-4630-bce8-755b7bb8492b.png)

Whatever, maybe there is some major refactoring under way and those options will come back in a feature build? The best thing would be to reintroduce it with an OpenGL or OpenGL ES hardware renderer. :+1: That would be really GREAT! ;-)

**Comment by nicolasnoble** [comment]
Out of curiosity, could you try with a retail bios?

**Comment by lorn10** [comment]
That was exactly my first thought. Unfortunately there is almost no difference.

I have tested this right now a further time (with European **SCPH-7502 BIOS v4.1**). While I have with the **PSX OpenBIOS** in **Crash Bandicoot** (at the slowest point, at the main menu screen) around 14.xx FPS I have with the retail bios around 16.xx FPS. That probably has no relevance, does it?

Whatever, maybe this will be better with the new **x86-64 recompiler**. ;-)

**Comment by wheremyfoodat** [comment]
@lorn10 14.xx and 16.xx FPS sounds like your PC is struggling to run Redux at fullspeed.
The x86-64 recompiler would help, yeah. I can send a build for you to try if you want to try it out without compiling it.

**Comment by lorn10** [comment]
Hi @wheremyfoodat, - thanks, that would be really great! :+1: 

However, as mentioned, this worked well with the old FPS limiter (in its final stage). Maybe this is really a Linux related problem. It would be interesting to check the Mac OS build. Will look into that when I find the time.

And regrading the hardware, - my test **iMac12,2** has a **Quad Core Intel Core i5-2500S CPU**, that should be enough I think. ;-P

```
test@test-iMac:~$ inxi -b
System:    Host: test-iMac Kernel: 5.11.0-38-generic x86_64 bits: 64 Desktop: KDE Plasma 5.18.5 
           Distro: Ubuntu 20.04.3 LTS (Focal Fossa) 
Machine:   Type: Desktop System: Apple product: iMac12,2 v: 1.0 serial: <superuser/root required> 
           Mobo: Apple model: Mac-942B59F58194171B v: iMac12,2 serial: <superuser/root required> UEFI: Apple 
           v: IM121.88Z.0047.B1F.1201241648 date: 01/24/12 
CPU:       Quad Core: Intel Core i5-2500S type: MCP speed: 1596 MHz min/max: 1600/3700 MHz 
Graphics:  Device-1: Intel 2nd Generation Core Processor Family Integrated Graphics driver: i915 v: kernel 
           Device-2: Advanced Micro Devices [AMD/ATI] Whistler [Radeon HD 6730M/6770M/7690M XT] driver: radeon v: kernel 
           Display: x11 server: X.Org 1.20.11 driver: modesetting,radeon FAILED: ati unloaded: fbdev,vesa 
           resolution: 2560x1440~60Hz 
           OpenGL: renderer: AMD TURKS (DRM 2.50.0 / 5.11.0-38-generic LLVM 12.0.1) 
           v: 3.3 Mesa 22.0.0-devel (git-68a6222 2021-10-30 focal-oibaf-ppa) 
Network:   Device-1: Broadcom and subsidiaries NetXtreme BCM57765 Gigabit Ethernet PCIe driver: tg3 
           Device-2: Qualcomm Atheros AR93xx Wireless Network Adapter driver: ath9k 
Drives:    Local Storage: total: 931.51 GiB used: 75.26 GiB (8.1%) 
Info:      Processes: 200 Uptime: 2d 1h 00m Memory: 7.75 GiB used: 1.76 GiB (22.7%) Shell: bash inxi: 3.0.38 
```

**Comment by wheremyfoodat** [comment]
Oh oof sorry, I can only build for Windows right now :(

**Comment by lorn10** [comment]
No problem, - will look into the Mac OS build. :+1: 

And additionally, maybe you can make an alternative Windows build with **Visual C++ 2015** or **Visual C++ 2017**? This may then run again in Wine. So I can test if this strange behavior is also present under **Linux/Wine**.

Little Background information, newer Wine versions since 6.3 have introduced additional built-in Visual C++ support like the "C2VectParallel function". This is great but it makes Wine now to some degree incompatible to the native Visual C++ 2019 libraries. So newer Wine versions do not run nicely and act strange when original Visual C++ 2019 is installed.

I have tested this up and down here: https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/issues/2161

**Comment by lorn10** [comment]
@nicolasnoble Any chance to set the **[Deployment Target](https://www.swiftdevjournal.com/supporting-older-versions-of-ios-and-macos/)** for the Mac OS build to a lower level like Mac OS El Capitan **10.11**? 

This would effectively allow to run PCSX-Redux (under Mac OS) down to the first **Aluminum iMac** model from 2007. ;-) As mentioned, the hardware specs should all be fulfilled. However, maybe there exists some limitations in the "building process" which prevents such a move?

@wheremyfoodat Any chance to build a **Visual C++ 2015** or **Visual C++ 2017** PCSX-Redux build (for testing in Wine)?

**Comment by nicolasnoble** [comment]
Visual C++ 2019 is required. 2017 or 2015 won't build the code. Wine should run the code properly granted the runtime is installed correctly. You may need to fix your wine installation and/or install the proper runtime.

We have no idea about how to build stuff on MacOS, really. Other people have contributed it for us, and it seems to be working for most people. Feel free to try and tweak the process and see what happens. It's controlled by those files here:

https://github.com/grumpycoders/pcsx-redux/blob/main/.github/workflows/macos-build.yml
https://github.com/grumpycoders/pcsx-redux/blob/main/.github/scripts/create-app.sh

But since the code requires a recent enough compiler and platform, I doubt this would work.

**Comment by lorn10** [comment]
Thanks for the informative answer. :+1:

Okay, in this case the Mac OS build stays for the moment compliant only to newer versions which come with support for **APFS**. Will try this out when I get access to a newer Apple computer. Maybe it is compatible to macOS Sierra and upwards.

In summary it can be said that Mac OS users of earlier Apple computers still have the alternative to install additionally a recent Linux distro. This would then allow to use PCSX-Redux also on vintage Apple hardware.

And regarding Visual C++ 2019, here is the incompatibility problem I mean. The native `ucrtbase.dll` is because of the mentioned issues (with newer wine versions) no longer included. This may be the reason for the "invalid string_view position" error (when PCSX-Redux is started). However, this will probably get better again when the built-in Visual C++ 2019 support of Wine is even more completed, - especially the currently missing `MSVCP140_ATOMIC_WAIT.dll`.
![pcsx-redux_wine_native_vcrun2019_ucrtbase_error](https://user-images.githubusercontent.com/9931771/140623684-25aaedc0-1a14-4937-bf1e-c6f30d0315c7.png)
![pcsx-redux_wine_native_vcrun2019_invalid_string_error](https://user-images.githubusercontent.com/9931771/140623689-de136c10-7b28-460e-996a-0765b2d1338c.png)

**Comment by nicolasnoble** [comment]
For the filesystem of the DMG file, I don't really know how hdiutil works, but this may be tweakable on this line here:

https://github.com/grumpycoders/pcsx-redux/blob/main/.github/workflows/macos-build.yml

For the missing dll, this is absolutely from the 2019 runtime redistributable from https://github.com/grumpycoders/pcsx-redux/blob/main/src/main/pcsx-wrapper.cc#L50-L51

**Comment by lorn10** [comment]
Yep, that may be the point. :+1: However, out of mine view, the `macOS-latest` argument strongly indicates that only newer Mac OS versions are supported...

Whatever, I am not a coder just a beta (or alpha?) tester. And if I don't really understand something, I keep my hands off it. ;-) Maybe someone who understands this matter better can make the necessary changes.

Back to the Visual C++ 2019 runtime matter. PCSX-Redux worked **almost perfectly well** in Wine with **built-in** Visual C++  2019. Then there was that change regarding the "Controls" aka "Input controller" config menu addition. And after that, I was no longer able to run PCSX-Redux in Wine. Maybe I will try these old builds out until I find the "last working one".

It's late now in Central Europe, - good night! :-P

**Comment by lorn10** [comment]
I can confirm, last working PCSX-Redux version (under Wine with **built-in** Visual C++ 2019) is **pcsx-redux-nightly-4299.20210608.6-x64**. 

The next build **pcsx-redux-nightly-4317.20210608.12-x64** does not start anymore. Interestingly, the error message is different (in comparison to the latest builds). It tells me:
```
0078:fixme:imm:ImeSetActiveContext (0x3c600, 0): stub
0078:fixme:imm:ImmReleaseContext (0000000000010020, 000000000003C600): stub
0110:fixme:imm:ImeSetActiveContext (0x36750, 1): stub
0110:fixme:imm:ImmReleaseContext (000000000001004C, 0000000000036750): stub
0100:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 000000000011FD70
```
While the newer builds gives me:
```
00d4:err:module:import_dll Library MSVCP140_ATOMIC_WAIT.dll (which is needed by L"Z:\\home\\test\\Downloads\\PCSX-Redux\\pcsx-redux.main") not found
00d4:err:module:LdrInitializeThunk Importing dlls for L"Z:\\home\\test\\Downloads\\PCSX-Redux\\pcsx-redux.main" failed, status c0000135
00cc:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 000000000011FD40
```

**Comment by wheremyfoodat** [comment]
I'm on the road right now so I can't check commit history but that seems to be a build from August 6th 2021. The bottom error definitely seems like an outdated Microsoft Visual C++ runtime.

Is there any way to update your MSVC++ redistributables on Wine from here? https://docs.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-160#visual-studio-2015-2017-2019-and-2022

**Comment by lorn10** [comment]
Hi wheremyfoodat! Well, as mentioned several times now, - Wine comes with **his own** MSVC++ implementation. Installing the original native Microsoft Visual C++ 2019 files is at the moment no solution. It will cause in Wine 6.x just more troubles...

However, usually it can be said that the built-in Visual C++ works **in Wine** better then the original one. This is true for almost all versions up to Visual C++ 2015. Visual C++ 2017 and 2019 are currently "work in progress".

**Comment by wheremyfoodat** [comment]
In that case there's not much to do aside from using the native builds, unless the wine team looks into it or some other workaround is found. I'll have a look at it eventually in a Manjaro VM.

**Comment by lorn10** [comment]
Okay, I have now looked more closely at the corresponding pulls which were merged at June 8, 2021. I would say, out of my "inner feeling", that pull https://github.com/grumpycoders/pcsx-redux/pull/505 may has a link to the broken Wine compatibility. This is just an assumption, but it would make sense in different ways.

However, the question is now if the "Windows wrapper for redistributables" pull is really needed? Or is it for example just an optional checking which could be also again excluded? Or maybe there was some further and larger architectural change between nightly-4299 and nightly-4317 which finally fundamentally depends on the MSVCP140_ATOMIC_WAIT.dll.

**Comment by nicolasnoble** [comment]
Remove the .exe and rename the .main file into a .exe.

**Comment by lorn10** [comment]
Well it looks that I was wrong, - the error still appears so it must be one of the other pulls. Nevertheless thanks for that hint, - I can now bypass the MSVC++ 2019 checking with that trick. :+1: 
```
0100:err:module:import_dll Library MSVCP140_ATOMIC_WAIT.dll (which is needed by L"Z:\\home\\test\\Downloads\\PCSX-Redux\\pcsx-redux.exe") not found
0100:err:module:LdrInitializeThunk Importing dlls for L"Z:\\home\\test\\Downloads\\PCSX-Redux\\pcsx-redux.exe" failed, status c0000135
```

**Comment by nicolasnoble** [comment]
It's likely less due to a specific commit, and more to a version update of visual studio.

**Comment by lorn10** [comment]
Aha yeah, that would be effectively also an explanation... And alternatively, also an additional "building option" or "building optimization" could trigger that, - whatever.

**Comment by lorn10** [comment]
Okay folks, good news, - for whatever reason **pcsx-redux-nightly-6141.20211105.2-x64** works now again under **Wine 6.21** with built-in MSVC++ 2019. :+1:  The **msvcp140_atomic_wait.dll** file is definitely needed, it has to be put into the PCSX-Redux directory. The previously mentioned "invalid string_view position error" no longer occurs under **Wine 6.21**. So it looks that there was some progress in that field.

Whatever, - unfortunately the situation regarding the **Speed Scaler** is under Wine even **worse**. The fluctuation of the emulation is bigger and in contrast to the native Linux build the entire system seems "heavier loaded" and lags. I don't think that I will be able to record a video in such a situation. ;-)

So my next target will be the Mac OS build. But I don't know when I get access to a newer Apple PC.
![pcsx-redux-nightly-6141_wine](https://user-images.githubusercontent.com/9931771/140935463-29f9b7ae-14fb-475b-9d4f-cb6273077e0c.png)

**Comment by wheremyfoodat** [comment]
Yeah, running under an intermediate layer like Wine should be expected to cause a performance hit, that might be significant.

What distro are you on? We might be able to get Redux building on your rig so you can try the x64 recompiler whenever you feel like it, until it's merged.

**Comment by lorn10** [comment]
I use **Kubuntu 20.04 LTS**, - but I can wait until the regular merge, no problem. ;-)

"Patience is half of life." (Maybe this wording does not fit in English, whatever.)

**Comment by wheremyfoodat** [comment]
In that case, you could try installing the required packages by doing `sudo apt-get install -y build-essential git make pkg-config clang g++ g++-mipsel-linux-gnu cpp-mipsel-linux-gnu binutils-mipsel-linux-gnu libfreetype-dev libavcodec-dev libavformat-dev libavutil-dev libglfw3-dev libswresample-dev libuv1-dev zlib1g-dev` in your terminal.

Then clone PCSX-Redux with submodules by doing `git clone <url> --recursive` (url = https://github.com/wheremyfoodat/pcsx-redux if you want to try out the new recompiler)
Then cd to the Redux root directory (the one with a Makefile) and type `make`

**Comment by lorn10** [comment]
A new status update. It looks that I was entirely on the wrong track, - this issue here **is not** speed scaler related. @nicolasnoble 

After some further testing I was able to figure out that Linux build **PCSX-Redux-6746a743-linux-x86_64** (October 02) is the effective last good working one. And that one already includes the "speed scaler" redesign, so it must be something other.

The next build **PCSX-Redux-3b28a1d6-linux-x86_64** (October 03) is the first with the "async audio/video speed slowdown behavior" which is present until the latest one. Maybe this helps now to find the effective root of the problem.

Whatever, sorry for the confusion. I was almost certainly sure that all this started much earlier with the introduction of the speed scaler.

**Comment by wheremyfoodat** [comment]
This is around the time when the callstack PR was merged, which seems to have caused some issues with Crash. We'll look into this a bit more.

**Comment by lorn10** [comment]
Additionally here follows now a small fun-fact, - the Windows build **pcsx-redux-nightly-5655.20211002.6-x64** (also from October 02) works under **Linux/Wine** absolutely equally good. And that at exactly the same frame rate of **FPS 50.xx**, - really no difference to the native Linux build! :+1:

But also here is a little intervention needed. The .exe file has to be removed/deleted and the .main file has to be renamed into a .exe. Otherwise it would not load. And of course, also the msvcp140_atomic_wait.dll file is needed, it must be present in the PCSX-Redux directory.

Whatever, really nice, I really didn't expect such a behavior for a "real emulator" under Wine. ;-) This shows that also a true emulator like PCSX-Redux can run under Wine quite good. So this is not only true for CXBX-R which is effectively more a Xbox "compatibility layer" rather than an emulator. (CXBX-R doesn't emulate the CPU, GPU and so on.) ;-)

**Comment by wheremyfoodat** [comment]
@lorn10 x86-64 dynarec is now merged into mainline Redux if you want to try it

**Comment by lorn10** [comment]
Thanks for the hint!

Okay it's gotten better and worse. ;-) The "async audio/video speed slowdown behavior"  is no longer present :+1:  but instead of that the screen output is massively corrupted. I made a little video of that:

https://drive.google.com/file/d/1GCi5HJWd8g7ukOMMUoWAeQzsYSSaL23_/view?usp=sharing

Whatever, maybe this shows now a new path to discover the underlying problem. I am back on build 6746a743. Did you check the changes in 3b28a1d6? As mentioned, for me on Linux the problem started there. ;-)

**Comment by lorn10** [comment]
Again a little addition, - the Windows **build 6411** runs under Linux/Wine almost error-free. That's somehow weird, but also interesting. Whatever, if you look very exactly at the screen output then you will notice still some very minor corruptions. And I think that also with the sound-flow there is something not fully correct.

For more details check the video here:
https://drive.google.com/file/d/1T983gbFj_tpXM2_jDuJ9O501ign6jv9B/view?usp=sharing

**Comment by lorn10** [comment]
Okay, the following may be something for @wheremyfoodat. 

In conjunction with **Crash Bandicoot Warped** I get with the new x86-64 dynarec and **Linux build** 45ebec0f an "**Unrecoverable error while running recompiler, Program counter: 8003A950**". It should be noted that this issue is most likely present also at the previous build but I didn't test it before. (The normal interpreter works but with the known "async audio/video speed slowdown behavior".)
![Crash_Bandicoot_Warped_PCSX-Redux_build45ebec0f](https://user-images.githubusercontent.com/9931771/144755594-c258cb6e-79bb-4d16-83d1-549e1499b424.png)

Interestingly this doesn't happen at the equivalent **Windows build 6438**. That runs in direct comparison quite good, apart from the mentioned small glitches. :
![Crash_Bandicoot_Warped_PCSX-Redux_build6438](https://user-images.githubusercontent.com/9931771/144755743-ebbe443c-2857-4aac-822f-42f4afaab668.png)

So it can be really said, at the moment the Windows build runs under Linux/Wine for whatever reason far better then the native Linux one. ;-)

**Comment by wheremyfoodat** [comment]
@lorn10 Gonna look into it in a bit. What part of the game does the unrecoverable error appear in? Before the title screen.

Btw it's probably best to mention stuff like this in separate issues so they don't get drowned here.

**Comment by wheremyfoodat** [comment]
![image](https://user-images.githubusercontent.com/44909372/144757873-3f521162-594a-4b4a-a032-957057d5e966.png)
![image](https://user-images.githubusercontent.com/44909372/144757919-074d72e1-cbd7-4a7c-8657-75c1f9bb6253.png)
@lorn10 
Works fine for me on Linux with a retail BIOS, though OpenBIOS seems to give me a recompiler error. Both interp and dynarec have some visual artifacting which is probably caused by either GPU or GTE bugs.

**Comment by lorn10** [comment]
Thanks for your super fast replay. Yes, the mentioned "unrecoverable error" appears after the OpenBIOS intro music so when the ROM is loaded.

But I can also confirm, it really works with the retail BIOS! :+1: , - in my case **SCPH-7502 BIOS v4.1**. But the "async audio/video speed problem" is still there when the normal interpreter is used. I have made a picture (right after the return from the first demo level) where you can see a **FPS** of **66.45**, so the sound is playing clearly too fast:
![Crash_Bandicoot_Warped_PCSX-Redux_build6438_RetailBIOS](https://user-images.githubusercontent.com/9931771/144759906-369774d1-3748-4976-b47d-16c2a9bb81a1.png)

With the DynaRec I have still some minor screen issues. Whatever, it looks that we have now a new bug related to the **PSX OpenBIOS** in conjunction with the new DynaRec. ;-)

**Comment by wheremyfoodat** [comment]
> Thanks for your super fast reply.
No probs

> With the DynaRec I have still some minor screen issues. 
Yeah, you can notice similar issues in Spyro too (at least the first game). And the PS logo in the BIOS bootup sequence also has some noticeable GPU artifacting. And so on, and so on. Fixing that would be a pretty gruesome task right now, but hopefully it might get done in the foreseeable future.

**Comment by lorn10** [comment]
Yes, this is still all quite "work in progress", - keep up that awesome motivation also in the new year! :+1: 

I think the new DyanRec is definitely a BIG improvement :1st_place_medal:,  especially for somewhat weaker x86-64 based systems. So the limited CPU power can be used even more efficient... 

And regarding Linux/Wine, the performance is thanks to the new DynaRec also on 64bit systems effectively on par with the native Linux build. I have encountered really no difference in the overall behavior. Just the remaining minor screen artifacts are somewhat annoying. But I am optimistic, - this "async audio/video speed issue" here will be fixed in the upcoming year. ;-)

**Comment by wheremyfoodat** [comment]
> Yes, this is still all quite "work in progress", - keep up that awesome motivation also in the new year! 👍
> 
> I think the new DyanRec is definitely a BIG improvement 🥇, especially for somewhat weaker x86-64 based systems. So the limited CPU power can be used even more efficient...
> 
> And regarding Linux/Wine, the performance is thanks to the new DynaRec also on 64bit systems effectively on par with the native Linux build. I have encountered really no difference in the overall behavior. Just the remaining minor screen artifacts are somewhat annoying. But I am optimistic, - this "async audio/video speed issue" here will be fixed in the upcoming year. ;-)

Hah, thanks, it means a lot to hear that. Gonna try to make it a tad faster too in the future, got some ideas though they're a little bit hard to implement at the time being, and there's some other things that have to be done too.

**Comment by lorn10** [comment]
A **GREAT** little news, - it's all fixed in Linux build **72dbece7** and Windows build **6489**. :+1: 

So no longer any "async audio/video speed issue" and also the previously mentioned "PSX OpenBIOS DynaRec crash" in Crash Bandicoot Warped is gone. Additionally, the speed of the normal interpreter is again much better, - even for the Windows build under Linux/Wine.

Really **great job** who ever has fixed this! :1st_place_medal: Out of mine view there are two merge candidates which come into question but I guess it was the "rewrite of the register allocator". ;-) 

So the only remaining issue for me are the minor "screen artifacts" when the new 64bit dynarec interpreter is used. But this is already known and a different thing. So I will close this issue here.

**Comment by wheremyfoodat** [comment]
> Here we go: https://drive.google.com/file/d/1GOdxoa7xjV_OBf3_lO9fm3HvLeJvZIjA/view?usp=sharing
> 
> It's recorded with **Kazam**, - the sound quality is not perfect and reduced. The "crackling" is because of the recording, this doesn't happen under normal conditions. As mentioned, the emulation is performed via the **Radeon dGPU**.
> 
> **Note**, the graphical glitches are "normal" for PCSX-Redux, these are there in **Crash Bandicoot Warped** since I have started to test this emulator. The original Crash Bandicoot game don't shows any such artifacts. However, this doesn't occurs at the old PCSX-Reloaded builds from the official Ubuntu PPA (tested version 1.9.92).
> 
> **Addition**, - the audio/video sync behavior in **Crash Bandicoot** is even worse: https://drive.google.com/file/d/1jYM2pqt5f_80rg8ng_f4xnDjHEtDDZWG/view?usp=sharing
> 
> Maybe this here is only Linux related? Whatever, if this is not reproducible with the Windows build then this is definitely a Linux related thing. Unfortunately I cannot test the Windows build in Wine anymore because it is compiled "more demanding":
> 
> ```
> 00d4:err:module:import_dll Library MSVCP140_ATOMIC_WAIT.dll (which is needed by L"Z:\\home\\test\\Downloads\\PCSX-Redux\\pcsx-redux.main") not found
> 00d4:err:module:LdrInitializeThunk Importing dlls for L"Z:\\home\\test\\Downloads\\PCSX-Redux\\pcsx-redux.main" failed, status c0000135
> 00cc:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 000000000011FD40
> ```
> 
> So it would be great if someone can compile a Visual C++ 2015 or 2017 PCSX-Redux "test build" which may then work also under Wine. ;-)

@lorn10 This is an ancient issue, but the graphical issues in Crash Bandicoot are fixed in #744
